core/DCache.cpp

```diff
@@ -35,71 +66,246 @@ namespace olympia {
+        const auto & mshr_itb = mem_access_info_ptr->getMSHRInfoIterator();
+
+        if (!mshr_itb.isValid() && mshr_file_.numFree() == 0)
+        {
+            // Should be Nack but miss should work for now
+            mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::MISS);
+            out_lsu_lookup_ack_.send(mem_access_info_ptr);
+            return;
+        }
+
+        if (!mshr_itb.isValid())
+        {
+            //            mshrLookup_(mem_access_info_ptr);
+            if (!mem_access_info_ptr->getMSHRInfoIterator().isValid())
+            {
+                ILOG("Creating new MSHR Entry " << mem_access_info_ptr);
+                allocateMSHREntry_(mem_access_info_ptr);
             }
         }
+
+        replyLSU(mem_access_info_ptr);
+    }
+
+    void DCache::replyLSU(const MemoryAccessInfoPtr & mem_access_info_ptr)
+    {
+        const auto & mshr_it = mem_access_info_ptr->getMSHRInfoIterator();
+        const uint64_t block_addr = getBlockAddr(mem_access_info_ptr);
+        const bool data_arrived = (*mshr_it)->getDataArrived();
+        const bool is_store_inst = mem_access_info_ptr->getInstPtr()->isStoreInst();
+
+        // All ST are considered Hit
+        if (is_store_inst)
+        {
+            // Update Line fill buffer only if ST
+            ILOG("Write to Line fill buffer (ST), block address:0x" << std::hex << block_addr);
+            (*mshr_it)->setModified(true);
+            (*mshr_it)->setMemRequest(mem_access_info_ptr);
+            mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::HIT);
+            out_lsu_lookup_ack_.send(mem_access_info_ptr);
+            return;
+        }
+
+        if (data_arrived)
+        {
+            ILOG("Hit on Line fill buffer (LD), block address:0x" << std::hex << block_addr);
+            mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::HIT);
+            out_lsu_lookup_ack_.send(mem_access_info_ptr);
+            return;
+        }
+
+        // Enqueue Load in LMQ
+        ILOG("Load miss inst to LMQ; block address:0x" << std::hex << block_addr);
+        (*mshr_it)->setMemRequest(mem_access_info_ptr);
+        mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::MISS);
+        out_lsu_lookup_req_.send(mem_access_info_ptr);
+    }
+
+    uint64_t DCache::getBlockAddr(const MemoryAccessInfoPtr & mem_access_info_ptr)
+    {
+        const InstPtr & inst_ptr = mem_access_info_ptr->getInstPtr();
+        const auto & inst_target_addr = inst_ptr->getRAdr();
+        return addr_decoder_->calcBlockAddr(inst_target_addr);
+    }
+
+    // Data read stage
+    void DCache::handleDataRead_()
+    {
+        const auto stage_id = static_cast<uint32_t>(PipelineStage::DATA_READ);
+        const MemoryAccessInfoPtr & mem_access_info_ptr = cache_pipeline_[stage_id];
+        ILOG(mem_access_info_ptr << " in read stage");
+        if (incoming_cache_refill_ == mem_access_info_ptr)
```

Comment:
If the MemoryAccessInfo keeps track of whether it's a refill or not, this code can change to:
```
if(mem_access_info_ptr->isRefill())
```